experimental/collator/src/provider.rs

```diff
@@ -0,0 +1,276 @@
+    const REORDERING_MASK: u32 = 1 << 6;
+    const LITHUANIAN_DOT_ABOVE_MASK: u32 = 1 << 7;
+    const BACWARD_SECOND_LEVEL_MASK: u32 = 1 << 8;
+    const ALTERNATE_SHIFTED_MASK: u32 = 1 << 9;
+    const CASE_FIRST_MASK: u32 = 1 << 10;
+    const UPPER_FIRST_MASK: u32 = 1 << 11;
+
+    #[inline(always)]
+    pub fn max_variable(&self) -> MaxVariable {
+        unsafe { core::mem::transmute((self.bits & CollationMetadataV1::MAX_VARIABLE_MASK) as u8) }
+    }
+
+    #[inline(always)]
+    pub fn numeric_primary(&self) -> u32 {
+        self.bits & 0xFF000000
+    }
+
+    #[inline(always)]
+    pub fn tailored(&self) -> bool {
+        self.bits & CollationMetadataV1::TAILORED_MASK != 0
+    }
+
+    /// Vietnamese and Ewe
+    #[inline(always)]
+    pub fn tailored_diacritics(&self) -> bool {
+        self.bits & CollationMetadataV1::TAILORED_DIACRITICS_MASK != 0
+    }
+
+    /// Korean search
+    #[inline(always)]
+    pub fn tailored_jamo(&self) -> bool {
+        self.bits & CollationMetadataV1::TAILORED_JAMO_MASK != 0
+    }
+
+    /// Lithuanian
+    #[inline(always)]
+    pub fn lithuanian_dot_above(&self) -> bool {
+        self.bits & CollationMetadataV1::LITHUANIAN_DOT_ABOVE_MASK != 0
+    }
+
+    /// Canadian French
+    #[inline(always)]
+    pub fn backward_second_level(&self) -> bool {
+        self.bits & CollationMetadataV1::BACWARD_SECOND_LEVEL_MASK != 0
+    }
+
+    #[inline(always)]
+    pub fn reordering(&self) -> bool {
+        self.bits & CollationMetadataV1::REORDERING_MASK != 0
+    }
+
+    /// Thai
+    #[inline(always)]
+    pub fn alternate_shifted(&self) -> bool {
+        self.bits & CollationMetadataV1::ALTERNATE_SHIFTED_MASK != 0
+    }
+
+    #[inline(always)]
+    pub fn case_first(&self) -> CaseFirst {
+        if self.bits & CollationMetadataV1::CASE_FIRST_MASK != 0 {
+            if self.bits & CollationMetadataV1::UPPER_FIRST_MASK != 0 {
+                CaseFirst::UpperFirst
+            } else {
+                CaseFirst::LowerFirst
+            }
+        } else {
+            CaseFirst::Off
+        }
+    }
+
+    // XXX alternate_non_ignorable Doesn't exist in current CLDR
+}
+
+#[icu_provider::data_struct(CollationSpecialPrimariesV1Marker = "collator/prim@1")]
+#[derive(Debug, PartialEq, Clone)]
+#[cfg_attr(feature = "datagen", derive(serde::Serialize))]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+pub struct CollationSpecialPrimariesV1<'data> {
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub last_primaries: ZeroVec<'data, u16>,
+    pub numeric_primary: u8,
+}
+
+impl<'data> CollationSpecialPrimariesV1<'data> {
+    pub(crate) fn last_primary_for_group(&self, max_variable: MaxVariable) -> u32 {
+        // Minus one to generate the right lower 16 bits from the high 16 bits.
+        // See parse.cpp in genrb and getLastPrimaryForGroup in ICU4C.
+        (u32::from(self.last_primaries.get(max_variable as usize).unwrap()) << 16) - 1
```

Comment:
Issue: Please either GIGO or validate, not unwrap (can be a follow-up)

Reply:
[Noted.](https://github.com/unicode-org/icu4x/issues/1891#issuecomment-1132667837)